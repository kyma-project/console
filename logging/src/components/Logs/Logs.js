import React, { createContext, useContext, useReducer, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Header from '../Header/Header';
import CompactHeader from '../CompactHeader/CompactHeader';
import LogTable from '../LogTable/LogTable';
import searchParamsReducer, {
  SET_LABELS,
  SET_SHOW_PREVIOUS_LOGS, SET_SEARCH_PHRASE,
  ADD_LABEL, SET_QUERY, SET_RESULT_LIMIT,
  SET_SHOW_ISTIO_LOGS, SET_SHOW_HEALTH_CHECKS,
  SET_AUTO_REFRESH, SET_LOGS_PERIOD, SET_SORT_DIR,
  SearchParamsContext
} from './SearchParams.reducer'
// import 'core-js/es/array/flat-map'; todo


import {
  SORT_ASCENDING,
  DEFAULT_PERIOD,
  LOG_REFRESH_INTERVAL,
} from './../../constants';

function sortLogs(entry1, entry2, sortDirection) {
  const positiveReturn = sortDirection === 'ascending' ? 1 : -1;
  const date1 = new Date(entry1.timestamp);
  const date2 = new Date(entry2.timestamp);
  return date1.getTime() > date2.getTime()
    ? positiveReturn
    : -1 * positiveReturn;
}

const Logs = ({ readonlyLabels, isCompact, httpService }) => {
  const defaultSearchParams = {
    searchPhrase: '',
    labels: [],
    readonlyLabels: readonlyLabels,
    logsPeriod: DEFAULT_PERIOD,
    query: '',
    resultLimit: 1000,
    showPreviousLogs: true,
    showHealthChecks: true,
    showIstioLogs: false,
    sortDirection: SORT_ASCENDING,
    logs: [],
    autoRefreshEnabled: true,
  };

  const [searchParams, dispatch] = useReducer(searchParamsReducer, defaultSearchParams);

  const [intervalId, setIntervalId] = useState(null);
  const [logs, setLogs] = useState([]);

  useEffect(() => {
    const { autoRefreshEnabled } = searchParams;
    if (intervalId) {
      clearInterval(intervalId)
    }

    if (autoRefreshEnabled) {
      startAutoRefresh(searchParams)
    }

    if (!autoRefreshEnabled && intervalId) {
      setIntervalId(null)
    }
  }, [searchParams])

  const filterHealthChecks = entry => {
    const { showHealthChecks } = searchParams;
    return showHealthChecks || !~entry.log.indexOf('GET /healthz');
  };

  async function fetchLogs(state) {
    const {
      searchPhrase,
      labels,
      readonlyLabels,
      sortDirection,
      logsPeriod,
      resultLimit,
      showPreviousLogs,
      showHealthChecks,
      showIstioLogs,
    } = state;

    if (!labels.length && !readonlyLabels.length && !searchPhrase) {
      return;
    }

    try {
      const result = await httpService.fetchLogs({
        searchPhrase,
        labels: [...readonlyLabels, labels],
        resultLimit,
        logsPeriod,
        sortDirection,
        showPreviousLogs,
        showHealthChecks,
      });
      console.log('fetchLogs')

      let streams = result.streams || [];

      if (!showIstioLogs) {
        streams = [...streams].filter(
          s => !~s.labels.indexOf('container_name="istio-proxy"'),
        );
      }
      const logs = streams
        .flatMap(stream => stream.entries)
        .map(l => ({
          timestamp: l.ts,
          log: l.line,
        }))
        .sort((e1, e2) => sortLogs(e1, e2, sortDirection));

      setLogs(logs)
    } catch (e) {
      console.warn(e); // todo add error message
    }
  };

  function startAutoRefresh(state) {
    fetchLogs(state);
    setIntervalId(setInterval(() => fetchLogs(state), LOG_REFRESH_INTERVAL))
  }
  const actions = {
    addLabel: label => dispatch({ type: ADD_LABEL, value: label }),
    setLabels: labels => dispatch({ type: SET_LABELS, value: labels }),
    setShowPreviousLogs: show => dispatch({ type: SET_SHOW_PREVIOUS_LOGS, value: show }),
    setShowHealthChecks: show => dispatch({ type: SET_SHOW_HEALTH_CHECKS, value: show }),
    setShowIstioLogs: show => dispatch({ type: SET_SHOW_ISTIO_LOGS, value: show }),
    setSearchPhrase: phrase => dispatch({ type: SET_SEARCH_PHRASE, value: phrase }),
    setQuery: query => dispatch({ type: SET_QUERY, value: query }),
    setResultLimit: limit => dispatch({ type: SET_RESULT_LIMIT, value: limit }),
    setAutoRefresh: isRefreshEnabled => dispatch({ type: SET_AUTO_REFRESH, value: isRefreshEnabled }),
    setLogsPeriod: period => dispatch({ type: SET_LOGS_PERIOD, value: period }),
    setSortDir: dir => dispatch({ type: SET_SORT_DIR, value: dir })
  };

  return (

    <SearchParamsContext.Provider value={[searchParams, actions]}>
      {isCompact ? (
        // <CompactHeader
        //   updateFilteringState={this.updateState}
        //   searchPhrase={searchPhrase}
        //   logsPeriod={logsPeriod}
        //   sortDirection={sortDirection}
        //   advancedSettings={advancedSettings}
        //   autoRefreshEnabled={autoRefreshEnabled}
        // />
        null
      ) : (
          <Header />
        )}

      <LogTable entries={logs.filter(filterHealthChecks)} />
    </SearchParamsContext.Provider>
  );

}

export default Logs;