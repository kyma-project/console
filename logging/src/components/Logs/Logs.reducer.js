import {createContext, useContext} from "react";

const LogsContext = createContext({});
export const useLogs = useContext(LogsContext);

export const SET_LABELS = 'SET_LABELS';
export const SET_SHOW_PREVIOUS_LOGS = 'SET_SHOW_PREVIOUS_LOGS';

function reducer(state, action) {
  switch (action.type) {
    case SET_LABELS: return {...state, labels: action.value};
    case SET_SHOW_PREVIOUS_LOGS: return {...state, showPreviousLogs: action.value};
    default: return state
  }
}