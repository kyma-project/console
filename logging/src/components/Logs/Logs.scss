.logs__no-filter {
    text-align: center;
}