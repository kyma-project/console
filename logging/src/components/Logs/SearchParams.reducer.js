import { createContext, useContext } from "react";
import {
  SORT_DESCENDING,
  DEFAULT_PERIOD
} from '../../constants'

export const SearchParamsContext = createContext({});

export const useSearchParams=()=>useContext(SearchParamsContext);

export const ADD_LABEL = 'ADD_LABEL';
export const SET_LABELS = 'SET_LABELS';
export const SET_SHOW_PREVIOUS_LOGS = 'SET_SHOW_PREVIOUS_LOGS';
export const SET_SHOW_HEALTH_CHECKS = 'SET_SHOW_HEALTH_CHECKS';
export const SET_SHOW_ISTIO_LOGS = 'SET_SHOW_ISTIO_LOGS';
export const SET_SEARCH_PHRASE = 'SET_SEARCH_PHRASE';
export const SET_RESULT_LIMIT = 'SET_RESULT_LIMIT';
export const SET_AUTO_REFRESH = 'SET_AUTO_REFRESH';
export const SET_SORT_DIR = 'SET_SORT_DIR';
export const SET_LOGS_PERIOD = 'SET_LOGS_PERIOD';

export const defaultSearchParams = {
  searchPhrase: '',
  labels: [],
  readonlyLabels: [],
  logsPeriod: DEFAULT_PERIOD,
  resultLimit: 1000,
  showPreviousLogs: false,
  showHealthChecks: false,
  showIstioLogs: false,
  sortDirection: SORT_DESCENDING,
  autoRefreshEnabled: true,
};

export const actions = (dispatch) => ({
  addLabel: label => dispatch({ type: ADD_LABEL, value: label }),
  setLabels: labels => dispatch({ type: SET_LABELS, value: labels }),
  setShowPreviousLogs: show => dispatch({ type: SET_SHOW_PREVIOUS_LOGS, value: show }),
  setShowHealthChecks: show => dispatch({ type: SET_SHOW_HEALTH_CHECKS, value: show }),
  setShowIstioLogs: show => dispatch({ type: SET_SHOW_ISTIO_LOGS, value: show }),
  setSearchPhrase: phrase => dispatch({ type: SET_SEARCH_PHRASE, value: phrase }),
  setResultLimit: limit => dispatch({ type: SET_RESULT_LIMIT, value: limit }),
  setAutoRefresh: isRefreshEnabled => dispatch({ type: SET_AUTO_REFRESH, value: isRefreshEnabled }),
  setLogsPeriod: period => dispatch({ type: SET_LOGS_PERIOD, value: period }),
  setSortDir: dir => dispatch({ type: SET_SORT_DIR, value: dir })
});


export default function searchParamsReducer(state, action) {
  const { type, value } = action;

  switch (type) {
    case ADD_LABEL: return { ...state, labels: [...state.labels, value] }
    case SET_LABELS: return { ...state, labels: value };

    case SET_SHOW_PREVIOUS_LOGS: return { ...state, showPreviousLogs: value };
    case SET_SHOW_HEALTH_CHECKS: return { ...state, showHealthChecks: value }
    case SET_SHOW_ISTIO_LOGS: return { ...state, showIstioLogs: value }

    case SET_SEARCH_PHRASE: return { ...state, searchPhrase: value }
    case SET_RESULT_LIMIT: return { ...state, resultLimit: value }

    case SET_AUTO_REFRESH: return { ...state, autoRefreshEnabled: value }

    case SET_SORT_DIR: return { ...state, sortDirection: value }
    case SET_LOGS_PERIOD: return { ...state, logsPeriod: value }
    default: return state
  }
}
