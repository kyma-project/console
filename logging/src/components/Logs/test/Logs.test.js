import React from 'react';
import renderer from 'react-test-renderer';
import Logs from '../Logs';
import { isTest } from 'apollo-utilities';

describe('Logs', () => {
  it('Renders with minimal props', () => {
    console.error = jest.fn();

    afterEach(() => {
      console.error.mockReset();
    });

    afterAll(() => {
      expect(console.error.mock.calls[0][0]).toMatchSnapshot(); // unique "key" prop warning
    });

    const component = renderer.create(
      <Logs httpService={{}} queryTransformService={{ toQuery: () => {} }} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders with lambda view props', () => {
    console.error = jest.fn();

    afterEach(() => {
      console.error.mockReset();
    });

    afterAll(() => {
      expect(console.error.mock.calls[0][0]).toMatchSnapshot(); // unique "key" prop warning
    });

    const component = renderer.create(
      <Logs
        httpService={{}}
        queryTransformService={{ toQuery: () => {} }}
        isLambda={true}
        readonlyLabels={[]}
        lambdaName="test"
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test.todo('Renders with some labels')
});
