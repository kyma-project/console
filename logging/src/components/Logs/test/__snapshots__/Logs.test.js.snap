// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logs Renders with lambda view props 1`] = `
Array [
  <div
    className="fd-panel fd-has-padding-regular fd-has-padding-bottom-none"
  >
    <h1
      className="fd-has-type-3 fd-has-padding-bottom-tiny"
    >
      Logs
    </h1>
    <section
      className="header__settings-group"
    >
      <section
        className="fd-display-l-inline-block"
      >
        <span
          className="caption-muted"
        >
          Labels
        </span>
        <div
          className="fd-combobox-input"
        >
          <div
            className="fd-popover"
          >
            <div
              onFocus={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="fd-popover__control"
              >
                <div
                  className="fd-combobox-control"
                  onClick={[Function]}
                >
                  <div
                    className="fd-input-group fd-input-group--after"
                  >
                    <input
                      className="fd-input"
                      placeholder="Select Label"
                      type="text"
                    />
                    <span
                      className="fd-input-group__addon fd-input-group__addon--after fd-input-group__addon--button"
                    >
                      <button
                        className=" fd-button--light sap-icon--navigation-down-arrow"
                      />
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </section>
      <section
        className="fd-has-margin-right-small"
      >
        <span
          className="caption-muted search-input__caption"
        >
          Search
          <div
            className="small-inline-help-wrapper"
          >
            <span
              className="fd-inline-help"
            >
              <span
                className="fd-inline-help__content fd-inline-help__content--right"
              >
                Search for logs by text (optional)
              </span>
            </span>
          </div>
        </span>
        <input
          className=""
          id="search-input"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
      </section>
      <span
        className="link-button fd-has-type-minus-1 header__settings-group__toggle"
        data-test-id="advanced-settings-toggle"
        onClick={[Function]}
        role="button"
      >
        Show Advanced Settings
      </span>
    </section>
    <div>
      <section
        className="labels-display"
      >
        <ul />
      </section>
      <div
        className="header__options-wrapper"
      >
        <button
          className="fd-button--light sap-icon--media-pause link-button fd-has-margin-right-tiny fd-has-type-minus-1"
          data-test-id="auto-refresh-button"
          onClick={[Function]}
          size="xs"
        >
          auto refresh
        </button>
        <span
          className="link-button fd-has-type-minus-1 select-dropdown"
        >
          <div
            className="fd-popover"
          >
            <div
              onFocus={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="fd-popover__control"
              >
                <button
                  className="fd-button--light sap-icon--past fd-has-margin-right-tiny"
                  onClick={[Function]}
                  size="xs"
                >
                  last 15 minutes
                </button>
              </div>
            </div>
          </div>
        </span>
        <span
          className="link-button fd-has-type-minus-1 select-dropdown"
        >
          <div
            className="fd-popover"
          >
            <div
              onFocus={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="fd-popover__control"
              >
                <button
                  className="fd-button--light sap-icon--sort fd-has-margin-right-tiny"
                  onClick={[Function]}
                  size="xs"
                >
                  ascending
                </button>
              </div>
            </div>
          </div>
        </span>
      </div>
    </div>
  </div>,
  <article
    className="fd-container fd-container--centered"
  >
    <p
      className="fd-has-type-5 fd-has-margin-large"
    >
      Add some labels to filter to see the logs
    </p>
  </article>,
]
`;

exports[`Logs Renders with minimal props 1`] = `
Array [
  <div
    className="fd-panel fd-has-padding-regular fd-has-padding-bottom-none"
  >
    <h1
      className="fd-has-type-3 fd-has-padding-bottom-tiny"
    >
      Logs
    </h1>
    <section
      className="header__settings-group"
    >
      <section
        className="fd-display-l-inline-block"
      >
        <span
          className="caption-muted"
        >
          Labels
        </span>
        <div
          className="fd-combobox-input"
        >
          <div
            className="fd-popover"
          >
            <div
              onFocus={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="fd-popover__control"
              >
                <div
                  className="fd-combobox-control"
                  onClick={[Function]}
                >
                  <div
                    className="fd-input-group fd-input-group--after"
                  >
                    <input
                      className="fd-input"
                      placeholder="Select Label"
                      type="text"
                    />
                    <span
                      className="fd-input-group__addon fd-input-group__addon--after fd-input-group__addon--button"
                    >
                      <button
                        className=" fd-button--light sap-icon--navigation-down-arrow"
                      />
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </section>
      <section
        className="fd-has-margin-right-small"
      >
        <span
          className="caption-muted search-input__caption"
        >
          Search
          <div
            className="small-inline-help-wrapper"
          >
            <span
              className="fd-inline-help"
            >
              <span
                className="fd-inline-help__content fd-inline-help__content--right"
              >
                Search for logs by text (optional)
              </span>
            </span>
          </div>
        </span>
        <input
          className=""
          id="search-input"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
      </section>
      <span
        className="link-button fd-has-type-minus-1 header__settings-group__toggle"
        data-test-id="advanced-settings-toggle"
        onClick={[Function]}
        role="button"
      >
        Show Advanced Settings
      </span>
    </section>
    <div>
      <section
        className="labels-display"
      >
        <ul />
      </section>
      <div
        className="header__options-wrapper"
      >
        <button
          className="fd-button--light sap-icon--media-pause link-button fd-has-margin-right-tiny fd-has-type-minus-1"
          data-test-id="auto-refresh-button"
          onClick={[Function]}
          size="xs"
        >
          auto refresh
        </button>
        <span
          className="link-button fd-has-type-minus-1 select-dropdown"
        >
          <div
            className="fd-popover"
          >
            <div
              onFocus={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="fd-popover__control"
              >
                <button
                  className="fd-button--light sap-icon--past fd-has-margin-right-tiny"
                  onClick={[Function]}
                  size="xs"
                >
                  last 15 minutes
                </button>
              </div>
            </div>
          </div>
        </span>
        <span
          className="link-button fd-has-type-minus-1 select-dropdown"
        >
          <div
            className="fd-popover"
          >
            <div
              onFocus={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="fd-popover__control"
              >
                <button
                  className="fd-button--light sap-icon--sort fd-has-margin-right-tiny"
                  onClick={[Function]}
                  size="xs"
                >
                  ascending
                </button>
              </div>
            </div>
          </div>
        </span>
      </div>
    </div>
  </div>,
  <article
    className="fd-container fd-container--centered"
  >
    <p
      className="fd-has-type-5 fd-has-margin-large"
    >
      Add some labels to filter to see the logs
    </p>
  </article>,
]
`;
